pkg/operator/upgradebackupcontroller/upgradebackupcontroller.go

```diff
@@ -0,0 +1,258 @@
+	}
+
+	if len(errs) > 0 {
+		return "", utilerrors.NewAggregate(errs)
+	}
+
+	return operatorv1.ConditionUnknown, nil
+}
+
+func manageRecoveryPod(ctx context.Context, nodeName, targetImagePullSpec, clusterBackupVersion string, client coreclientv1.PodsGetter, recorder events.Recorder) (*corev1.Pod, bool, error) {
+	pod := resourceread.ReadPodV1OrDie(etcd_assets.MustAsset("etcd/cluster-backup-pod.yaml"))
+	pod.Name = "cluster-backup-" + nodeName
+	pod.Namespace = operatorclient.TargetNamespace
+	pod.Spec.NodeName = nodeName
+	pod.Labels = map[string]string{
+		"app":                     "cluster-backup",
+		clusterBackupVersionLabel: clusterBackupVersion,
+	}
+	pod.Spec.Containers[0].Image = targetImagePullSpec
+
+	return resourceapply.ApplyPod(ctx, client, recorder, pod)
+}
+
+// isRequireRecentBackup checks the conditions of ClusterVersion to verify if a backup is required.
+func isRequireRecentBackup(config *configv1.ClusterVersion) bool {
+	for _, condition := range config.Status.Conditions {
+		// Check if Failing is true and Message field containers the string RecentBackup.
+		if condition.Type == "Failing" && condition.Status == configv1.ConditionTrue {
+			return strings.Contains(condition.Message, "RecentBackup") == true
+		}
+	}
+	return false
+}
+
+// getNodeWithAdequateAllocatableStorage returns a control-plane node that meet the minimum storage requirements for the backup state.
+func getNodeWithAdequateAllocatableStorage(_ context.Context, podLister corev1listers.PodLister, nodeLister corev1listers.NodeLister) (*corev1.Node, error) {
+	pods, err := podLister.List(labels.Set{"app": "etcd"}.AsSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pod := range pods {
+		if !strings.HasPrefix(pod.Name, "etcd-") {
+			continue
+		}
+		node, err := nodeLister.Get(pod.Spec.NodeName)
+		if err != nil {
+			return nil, err
+		}
+
+		// Ensure nodes have adequate storage for backup state.
+		if !isNodeAllocatableStorageAdequate(node) {
+			klog.V(4).Infof("node storage is not adequate for backup: %s, allocatable %s, required: %d", node.Name, node.Status.Allocatable.Storage().String(), etcdenvvar.QuotaBackendBytes*2)
+			continue
+		}
+		return node, nil
+	}
+
+	return nil, fmt.Errorf("no control-plane nodes found with adequate storage for backup")
+}
+
+// isNodeAllocatableStorageAdequate checks the nodes reported allocatable storage. Returns true if greater than
+// QuotaBackendBytes x 2.
+func isNodeAllocatableStorageAdequate(node *corev1.Node) bool {
+	if node == nil {
+		return false
+	}
+	nodeAllocatableStorageBytes, ok := node.Status.Allocatable.Storage().AsInt64()
+	if !ok {
+		klog.V(4).Infof("error converting allocatable storage to int64 %v", node.Status.Allocatable.Storage())
+		return false
+	}
+	// Require 2x worst case backup size as allocatable storage
+	// TODO is this too extreme? We can ask end point status for the actual size of the snapshot at the expense of a 2 RPC calls.
```

Comment:
I would agree to do the calls here instead. We only do this operation on upgrades and if certain conditions are met so we won't be doing that many calls?

Reply:
OK more time but I want this right vs done.